genesis/sensors/base_sensor.py

```diff
@@ -12,10 +21,127 @@ class Sensor(RBC):
     A sensor must have a read() method that returns the sensor data.
     """
 
+    def __init__(self, sensor_options: "SensorOptions", sensor_idx: int, sensor_manager: "SensorManager"):
+        self._options: "SensorOptions" = sensor_options
+        self._idx: int = sensor_idx
+        self._manager: "SensorManager" = sensor_manager
+        self._cache_idx: int = -1  # cache_idx is set by the SensorManager during the scene build phase
+        self._read_delay_steps: int = 0
+        self._cache_buffer_length: int = 0
+
+    # =============================== implementable methods ===============================
+
+    @gs.assert_unbuilt
+    def build(self):
+        """
+        This method is called by the SensorManager during the scene build phase to initialize the sensor.
+        This is where any shared metadata should be initialized.
+        """
+        delay_steps_float = self._options.read_delay / self._manager._sim.dt
+        self._read_delay_steps = round(delay_steps_float)
+        if not np.isclose(delay_steps_float, self._read_delay_steps):
+            gs.logger.warn(
+                f"Read delay should be a multiple of the simulation time step. Got {self._options.read_delay} and "
+                f"{self._manager._sim.dt}. Actual read delay will be {1/self._read_delay_steps}."
+            )
+        self._cache_buffer_length = self._read_delay_steps + 1
```

Comment:
I don' think it is the right way to do it. The cache buffer length should be the same for all sensors, and the sensor manager should use `read_delay_step = max(sensor._read_delay_steps for sensor in self._sensors)` to determine how many data mush be stored in cache. It will make everything much easier to implement and faster, but increase the memory footprint to some extent.